readthedocs/core/symlink.py

```diff
@@ -0,0 +1,285 @@
+
+        # Remove old symlinks
+        if os.path.exists(self.subproject_root):
+            for subproj in os.listdir(self.subproject_root):
+                if subproj not in subprojects:
+                    os.unlink(os.path.join(self.subproject_root, subproj))
+
+    def symlink_translations(self):
+        """Symlink project translations
+
+        Link from $WEB_ROOT/<project>/<language>/ ->
+                  $WEB_ROOT/<translation>/<language>/
+        """
+        translations = {}
+
+        for trans in self.project.translations.all():
+            translations[trans.language] = trans.slug
+
+        # Make sure the language directory is a directory
+        language_dir = os.path.join(self.project_root, self.project.language)
+        if os.path.islink(language_dir):
+            os.unlink(language_dir)
+        if not os.path.exists(language_dir):
+            os.makedirs(language_dir)
+
+        for (language, slug) in translations.items():
+            self._log(u"Symlinking translation: {0}->{1}".format(language, slug))
+            symlink = os.path.join(self.project_root, language)
+            docs_dir = os.path.join(self.WEB_ROOT, slug, language)
+            print run('ln -nsf {0} {1}'.format(docs_dir, symlink))
+
+        # Remove old symlinks
+        for lang in os.listdir(self.project_root):
+            if (lang not in translations and
+                    lang not in ['projects', self.project.language]):
+                to_delete = os.path.join(self.project_root, lang)
+                if os.path.islink(to_delete):
+                    os.unlink(to_delete)
+                else:
+                    shutil.rmtree(to_delete)
+
+    def symlink_single_version(self):
+        """Symlink project single version
+
+        Link from $WEB_ROOT/<project> ->
+                  HOME/user_builds/<project>/rtd-builds/latest/
+        """
+        default_version = self.project.get_default_version()
+        self._log("Symlinking single_version")
+
+        symlink = self.project_root
+        if os.path.islink(symlink):
+            os.unlink(symlink)
+        if os.path.exists(symlink):
+            shutil.rmtree(symlink)
+
+        # Where the actual docs live
+        docs_dir = os.path.join(settings.DOCROOT, self.project.slug, 'rtd-builds', default_version)
+        print run('ln -nsf %s/ %s' % (docs_dir, symlink))
+
+    def symlink_versions(self):
+        """Symlink project's versions
+
+        Link from $WEB_ROOT/<project>/<language>/<version>/ ->
+                  HOME/user_builds/<project>/rtd-builds/<version>
+        """
+        versions = set()
+        version_dir = os.path.join(self.WEB_ROOT, self.project.slug, self.project.language)
+        # Include active public versions,
+        # as well as public verisons that are built but not active, for archived versions
+        version_queryset = (self.project.versions.public(only_active=False).filter(built=True) |
```

Comment:
This is because we don't pass in kwargs to public() -- https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/privacy/backend.py#L81 -- this was removed because we were passing in things that weren't getting respected properly.